File: app/src/main/java/org/ole/planet/myplanet/ui/course/CourseDetailFragment.java
package org.ole.planet.myplanet.ui.course;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.gson.JsonObject;

import org.ole.planet.myplanet.MainApplication;
import org.ole.planet.myplanet.base.BaseContainerFragment;
import org.ole.planet.myplanet.callback.OnRatingChangeListener;
import org.ole.planet.myplanet.databinding.FragmentCourseDetailBinding;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.model.RealmCourseStep;
import org.ole.planet.myplanet.model.RealmMyCourse;
import org.ole.planet.myplanet.model.RealmMyLibrary;
import org.ole.planet.myplanet.model.RealmRating;
import org.ole.planet.myplanet.model.RealmStepExam;
import org.ole.planet.myplanet.model.RealmUserModel;
import org.ole.planet.myplanet.service.UserProfileDbHandler;
import org.ole.planet.myplanet.utilities.Markdown;

import java.util.List;

import io.realm.Realm;
import io.realm.RealmResults;

public class CourseDetailFragment extends BaseContainerFragment implements OnRatingChangeListener {
    private FragmentCourseDetailBinding fragmentCourseDetailBinding;
    DatabaseService dbService;
    Realm mRealm;
    RealmMyCourse courses;
    RealmUserModel user;
    String id;

    public CourseDetailFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            id = getArguments().getString("courseId");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        fragmentCourseDetailBinding = FragmentCourseDetailBinding.inflate(inflater, container, false);
        dbService = new DatabaseService(getActivity());
        mRealm = dbService.getRealmInstance();
        courses = mRealm.where(RealmMyCourse.class).equalTo("courseId", id).findFirst();
        user = new UserProfileDbHandler(getActivity()).getUserModel();
        return fragmentCourseDetailBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initRatingView("course", courses.courseId, courses.courseTitle, this);
        setCourseData();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    private void setCourseData() {
        setTextViewVisibility(fragmentCourseDetailBinding.subjectLevel, courses.subjectLevel, fragmentCourseDetailBinding.ltSubjectLevel);
        setTextViewVisibility(fragmentCourseDetailBinding.method, courses.method, fragmentCourseDetailBinding.ltMethod);
        setTextViewVisibility(fragmentCourseDetailBinding.gradeLevel, courses.gradeLevel, fragmentCourseDetailBinding.ltGradeLevel);
        setTextViewVisibility(fragmentCourseDetailBinding.language, courses.languageOfInstruction, fragmentCourseDetailBinding.ltLanguage);
        String markdownContentWithLocalPaths = CourseStepFragment.prependBaseUrlToImages(courses.description, "file://" + MainApplication.context.getExternalFilesDir(null) + "/ole/");
        Markdown.INSTANCE.setMarkdownText(fragmentCourseDetailBinding.description, markdownContentWithLocalPaths);
        fragmentCourseDetailBinding.noOfExams.setText(RealmStepExam.getNoOfExam(mRealm, id) + "");
        final RealmResults resources = mRealm.where(RealmMyLibrary.class).equalTo("courseId", id).equalTo("resourceOffline", false).isNotNull("resourceLocalAddress").findAll();
        setResourceButton(resources, fragmentCourseDetailBinding.btnResources);
        final List<RealmMyLibrary> downloadedResources = mRealm.where(RealmMyLibrary.class).equalTo("resourceOffline", true).equalTo("courseId", id).isNotNull("resourceLocalAddress").findAll();
        setOpenResourceButton(downloadedResources, fragmentCourseDetailBinding.btnOpen);
        onRatingChanged();
        setStepsList();
    }

    private void setTextViewVisibility(TextView textView, String content, View layout) {
        if (content.isEmpty()) {
            layout.setVisibility(View.GONE);
        } else {
            textView.setText(content);
        }
    }

    private void setStepsList() {
        List<RealmCourseStep> steps = RealmCourseStep.getSteps(mRealm, courses.courseId);
        fragmentCourseDetailBinding.stepsList.setLayoutManager(new LinearLayoutManager(getActivity()));
        fragmentCourseDetailBinding.stepsList.setAdapter(new AdapterSteps(getActivity(), steps, mRealm));
    }

    @Override
    public void onRatingChanged() {
        JsonObject object = RealmRating.getRatingsById(mRealm, "course", courses.courseId, user.id);
        setRatings(object);
    }

    @Override
    public void onDownloadComplete() {
        super.onDownloadComplete();
        setCourseData();
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/courses/CourseDetailFragment.kt
package org.ole.planet.myplanet.ui.courses

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.core.view.doOnPreDraw
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.model.LazyHeaders
import com.bumptech.glide.signature.ObjectKey
import dagger.hilt.android.AndroidEntryPoint
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.base.BaseContainerFragment
import org.ole.planet.myplanet.callback.OnRatingChangeListener
import org.ole.planet.myplanet.databinding.FragmentCourseDetailBinding
import org.ole.planet.myplanet.model.RealmMyCourse
import org.ole.planet.myplanet.model.StepItem
import org.ole.planet.myplanet.utils.MarkdownUtils.setMarkdownText
import org.ole.planet.myplanet.utils.UrlUtils
import org.ole.planet.myplanet.utils.collectWhenStarted

@AndroidEntryPoint
class CourseDetailFragment : BaseContainerFragment(), OnRatingChangeListener {
    private var _binding: FragmentCourseDetailBinding? = null
    private val binding get() = _binding!!
    private var id: String? = null
    private val viewModel: CourseDetailViewModel by viewModels()
    private var isRatingViewInitialized = false
    private var stepsAdapter: CoursesStepsAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            id = requireArguments().getString("courseId")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentCourseDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        id?.let {
            viewModel.loadCourseDetail(it)
        }

        collectWhenStarted(viewModel.uiState) { state ->
            when (state) {
                is CourseDetailUiState.Loading -> {
                    // Show loading indicator if needed
                }
                is CourseDetailUiState.Success -> {
                    bindCourseData(state)
                }
                is CourseDetailUiState.Error -> {
                    context?.let { ctx ->
                        Toast.makeText(ctx, state.message, Toast.LENGTH_LONG).show()
                    }
                }
            }
        }

        collectWhenStarted(viewModel.stepItems) { steps ->
            setStepsList(steps)
        }
    }

    private fun bindCourseData(state: CourseDetailUiState.Success) {
        val course = state.course
        setTextViewVisibility(binding.subjectLevel, course.subjectLevel, binding.ltSubjectLevel)
        setTextViewVisibility(binding.method, course.method, binding.ltMethod)
        setTextViewVisibility(binding.gradeLevel, course.gradeLevel, binding.ltGradeLevel)
        setTextViewVisibility(binding.language, course.languageOfInstruction, binding.ltLanguage)
        setCourseCover(course.courseId, course.coverFileName, course.courseRev)

        setMarkdownText(binding.description, state.markdownDescription)

        binding.noOfExams.text = context?.getString(
            R.string.number_placeholder,
            state.examCount
        )

        setResourceButton(state.resources, binding.btnResources)
        setOpenResourceButton(state.downloadedResources, binding.btnOpen)

        if (!isRatingViewInitialized) {
            initRatingView("course", course.courseId, course.courseTitle, this@CourseDetailFragment)
            isRatingViewInitialized = true
        }
        setRatings(state.ratingSummary)
    }

    private fun setCourseCover(courseId: String?, coverFileName: String?, courseRev: String?) {
        val coverFile = RealmMyCourse.getCoverImageFile(binding.courseCover.context, courseId, coverFileName)
        val model: Any? = if (coverFile?.exists() == true) {
            coverFile
        } else {
            UrlUtils.getCourseImageUrl(courseId, coverFileName)?.let { url ->
                GlideUrl(url, LazyHeaders.Builder().addHeader("Authorization", UrlUtils.header).build())
            }
        }
        if (model == null) {
            binding.courseCover.visibility = View.GONE
            return
        }
        binding.courseCover.visibility = View.VISIBLE
        Glide.with(binding.courseCover.context)
            .load(model)
            .diskCacheStrategy(DiskCacheStrategy.ALL)
            .signature(ObjectKey(courseRev ?: ""))
            .error(R.drawable.ole_logo)
            .into(binding.courseCover)
    }

    private fun setTextViewVisibility(textView: TextView, content: String?, layout: View) {
        if (content?.isEmpty() == true) {
            layout.visibility = View.GONE
        } else {
            layout.visibility = View.VISIBLE
            textView.text = content
        }
    }

    private fun setStepsList(steps: List<StepItem>) {
        if (stepsAdapter == null) {
            binding.stepsList.layoutManager = LinearLayoutManager(activity)
            stepsAdapter = CoursesStepsAdapter(requireActivity()) { stepId ->
                viewModel.toggleStepDescription(stepId)
            }
            binding.stepsList.adapter = stepsAdapter
        }
        stepsAdapter?.submitList(steps)
    }

    override fun onRatingChanged() {
        id?.let {
            viewModel.refreshRatings(it)
        }
    }

    override fun onDownloadComplete() {
        super.onDownloadComplete()
        id?.let {
            viewModel.loadCourseDetail(it)
        }
    }

    override fun onDestroyView() {
        _binding = null
        stepsAdapter = null
        super.onDestroyView()
    }
}
